packages/editor/src/components/rich-text/format-toolbar/index.js

```diff
@@ -22,176 +11,49 @@ import { FORMATTING_CONTROLS } from '../formatting-controls';
-		}
-	}
-
-	isFormatActive( format ) {
-		return this.props.formats[ format ] && this.props.formats[ format ].isActive;
-	}
-
-	render() {
-		const { formats, enabledControls = [], customControls = [], selectedNodeId } = this.props;
-		const { linkValue, settingsVisible, opensInNewWindow, isEditingLink } = this.state;
-		const isAddingLink = formats.link && formats.link.isAdding;
-		const isEditing = isAddingLink || isEditingLink;
-		const isPreviewing = ! isEditing && formats.link;
-
-		const toolbarControls = FORMATTING_CONTROLS.concat( customControls )
-			.filter( ( control ) => enabledControls.indexOf( control.format ) !== -1 )
-			.map( ( control ) => {
-				if ( control.format === 'link' ) {
-					const isFormatActive = this.isFormatActive( 'link' );
-					const isActive = isFormatActive || isAddingLink;
-					return {
-						...control,
-						icon: isFormatActive ? 'editor-unlink' : 'admin-links', // TODO: Need proper unlink icon
-						title: isFormatActive ? __( 'Unlink' ) : __( 'Link' ),
-						onClick: isActive ? this.dropLink : this.addLink,
-						isActive,
-					};
-				}
-
+const FormatToolbar = ( props ) => {
+	const link = props.getActiveFormat( 'a' );
+	const removeLink = () => props.removeFormat( 'a' );
+	const addLink = () => props.applyFormat( {
+		type: 'a',
+		attributes: {
+			href: '',
+		},
+	} );
+
+	const toolbarControls = FORMATTING_CONTROLS
+		.filter( ( control ) => props.enabledControls.indexOf( control.format ) !== -1 )
+		.map( ( control ) => {
+			if ( control.format === 'link' ) {
 				return {
 					...control,
-					onClick: this.toggleFormat( control.format ),
-					isActive: this.isFormatActive( control.format ),
+					icon: link ? 'editor-unlink' : 'admin-links', // TODO: Need proper unlink icon
+					title: link ? __( 'Unlink' ) : __( 'Link' ),
+					onClick: link ? removeLink : addLink,
+					isActive: !! link,
 				};
-			} );
-
-		return (
-			<ToolbarContainer>
-				<Toolbar controls={ toolbarControls } />
-
-				{ ( isEditing || isPreviewing ) && (
-					<LinkContainer
-						editLink={ this.editLink }
-						formats={ formats }
-						isEditing={ isEditing }
-						isPreviewing={ isPreviewing }
-						linkValue={ linkValue }
-						onChangeLinkValue={ this.onChangeLinkValue }
-						onKeyDown={ this.onKeyDown }
-						opensInNewWindow={ opensInNewWindow }
-						selectedNodeId={ selectedNodeId }
-						setLinkTarget={ this.setLinkTarget }
-						settingsVisible={ settingsVisible }
-						submitLink={ this.submitLink }
-						toggleLinkSettingsVisibility={ this.toggleLinkSettingsVisibility }
-					/>
-				) }
-			</ToolbarContainer>
-		);
-	}
-}
+			}
 
-export default withSpokenMessages( FormatToolbar );
+			return {
+				...control,
+				onClick: () => props.toggleFormat( { type: control.selector } ),
+				isActive: !! props.getActiveFormat( control.selector ),
+			};
+		} );
+
+	return (
+		<ToolbarContainer>
+			<Toolbar controls={ toolbarControls } />
+			{ link && <LinkContainer
```

Comment:
This was raised by @noisysocks but let me repeat. It would be nice if we could convert rendering `<LinkContainer />` into render function and include it inside formatter definintion as a `render` option:
```js
render: ( props ) =>  <LinkContainer ... >
```

Reply:
See https://github.com/WordPress/gutenberg/pull/7890#issuecomment-417007246, but I have some more work on this locally. Definitely my plan to do this afterwards.